shard.go

```diff
@@ -0,0 +1,138 @@
+package flatfs
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+type shardId struct {
+	version string
+	funName string
+	param   string
+}
+
+func (f shardId) str() string {
+	if f.funName == "" || f.funName == "auto" {
+		return "auto"
+	} else {
+		return fmt.Sprintf("/repo/flatfs/shard/v1/%s/%s", f.funName, f.param)
+	}
+}
+
+func parseShardFunc(str string) shardId {
+	str = strings.TrimSpace(str)
+	parts := strings.Split(str, "/")
+	// ignore prefix for now
+	if len(parts) > 3 {
+		parts = parts[len(parts)-3:]
+	}
+	switch len(parts) {
+	case 3:
+		return shardId{version: parts[0], funName: parts[1], param: parts[2]}
```

Comment:
Okay, I will see what I can do.  I originally had parseShardFunc return an error but decided to push all the error checking into Func().  Will try refactoring again and see if I can make it work better the way I think you want.